### ISSUE TITLE
Feat: Develop and Track a Baseline Model

### ISSUE THREAD

**Issue by zhu-weijie** [zhu-weijie/creator-pulse#13]
### Goal

To create a model training component that trains a simple baseline machine learning model. This component will handle feature extraction (TF-IDF), model training, and will connect to our remote MLflow server to log all experiment parameters, metrics, and the trained model artifact.

### Tasks

- [ ] Add `mlflow` and `python-dotenv` to `requirements.txt`.
- [ ] Update `docker-compose.yml` to pass environment variables from the `.env` file to the container.
- [ ] Create a new script: `src/components/model_trainer.py`.
- [ ] Implement logic to load the preprocessed data from the previous step.
- [ ] Perform feature extraction using `TfidfVectorizer`.
- [ ] Train a simple baseline model (e.g., Logistic Regression).
- [ ] Log the model's parameters (e.g., vectorizer settings) and metrics (e.g., accuracy) to the MLflow server.
- [ ] Save the trained model as an artifact in the MLflow run.

**Comment by zhu-weijie** [comment]
<img width="1180" alt="Image" src="https://github.com/user-attachments/assets/bfd4024a-0b79-4645-b7ad-90271ed2e003" />

<img width="1194" alt="Image" src="https://github.com/user-attachments/assets/69cda6a4-9ea3-4a90-bfe3-ec14c29d8332" />

**Comment by zhu-weijie** [comment]
## An MLOps Project Flow

Here is a detailed breakdown of how your Python code, running inside a Docker container on your machine, communicates with the MLflow services running on AWS.

### The High-Level Analogy

Think of it like a project manager (your Python script) with two specialized assistants (MLflow Server and S3):

1.  **The Note-Taker (MLflow Server on EC2):** This assistant is responsible for writing down all the small, important details of your experiment: who ran it, what parameters were used (`max_features=5000`), and what the final results were (`accuracy=0.87`). It keeps a neat, organized logbook.
2.  **The Warehouse Manager (Amazon S3 Bucket):** This assistant handles the large, bulky items. When you have a finished product (the trained model file, which can be large), you tell the warehouse manager to store it.

Your Python script never sends the large model file *to* the Note-Taker on EC2. It asks the Note-Taker *where* to put the model, and the Note-Taker points to a specific shelf in the S3 warehouse.

---

### The Step-by-Step Technical Workflow

This diagram shows the flow of information:

```
+---------------------------------+
| Your Machine (Docker Container) |
|    [model_trainer.py]           |
+---------------------------------+
     |           |           |
     | 1. Read   |           |
     |   .env    |           |
     ↓           |           |
+------------------+         |
| [MLFLOW_TRACKING_URI]      |
| [http://52.xx.xx.xx:5000]  |
+------------------+         |
     |                       |
     | 2. Connect & Log      |  5. Log Model Artifact
     |    Metadata           |     (Directly to S3)
     | (params, metrics)     |           |
     ↓                       ↓           ↓
+-----------------------+   +------------------------------+
| AWS Cloud             |   | AWS Cloud                    |
|                       |   |                              |
| [EC2 Instance]        |   | [S3 Bucket]                  |
|  - MLflow Server      |   |  - creatorpulse-mlflow-..    |
|  - Listens on port 5000 |   |  - Stores the model.pkl file |
+-----------------------+   +------------------------------+
           ↑
           | 3. Get Storage Location
           |    (pre-signed S3 URL)
           |
           | 4. Server provides URL
           |
           +-----------------------
```

Here's what happens in your `model_trainer.py` script:

**Step 1: Configuration (Knowing Who to Call)**

```python
load_dotenv()
self.mlflow_tracking_uri = os.getenv("MLFLOW_TRACKING_URI")
```

*   Your script first reads the `.env` file. This file acts as its address book.
*   It finds the line `MLFLOW_TRACKING_URI=http://52.221.239.36:5000`. This tells your script the public IP address and port of your "Note-Taker" (the MLflow server on EC2).

**Step 2: Establishing the Connection (The Handshake)**

```python
mlflow.set_tracking_uri(self.mlflow_tracking_uri)
```

*   This is the crucial connection step. The `mlflow` library in your Python script opens a standard HTTP network connection to the specified address.
*   From now on, any `mlflow` command will send its information over the internet to your EC2 instance.

**Step 3: Logging Metadata (Sending Small Notes)**

```python
with mlflow.start_run(run_name="Baseline_LogisticRegression"):
    mlflow.log_param("max_features", 5000)
    mlflow.log_metric("accuracy", 0.869)
```

*   When this code runs, the `mlflow` library makes simple API calls to your server. It's essentially sending a request like:
    *   "Hi, `http://52.221.239.36:5000`! Please start a new run for me."
    *   "For the run you just created, please log a parameter: `max_features` is `5000`."
    *   "For the same run, please log a metric: `accuracy` is `0.869`."
*   The MLflow server on EC2 receives these requests and writes this small, structured data into its backend storage (which, by default, is a file-based system right on the EC2 instance's disk).

**Step 4: Logging Artifacts (Storing the Bulky Model File)**

This is the most important part to understand, involving both AWS services.

```python
mlflow.sklearn.log_model(sk_model=model, artifact_path="model", ...)
```

This command triggers a two-step process:

1.  **Your script asks the server for permission:** Your code sends a message to the MLflow server on EC2: "I have a model artifact I need to log for the current run. Where should I upload it?"
2.  **The server provides a one-time pass:** The MLflow server knows it was configured with a default artifact location (`--default-artifact-root s3://creatorpulse-mlflow-artifacts...`). It uses the AWS credentials on the EC2 instance to generate a special, temporary, secure upload link called a **pre-signed S3 URL**. It sends this unique URL back to your script.
3.  **Your script uploads directly to S3:** The `mlflow` library (using the `boto3` AWS library under the hood) takes your trained model file (`model.pkl`) and uploads it *directly* to that special S3 URL. The data **does not** pass through the EC2 server. This is highly efficient.

### Summary

The communication is cleverly split:

*   **All small metadata** (parameters, metrics, run names) is sent as simple API calls to the **MLflow Server on EC2**.
*   **All large data** (models, plots, data files) is sent directly from your code to the **S3 bucket**, using a temporary access URL provided by the EC2 server.

This architecture is scalable and efficient. The small EC2 instance only has to manage the metadata database, while the heavy lifting of file storage is offloaded to Amazon S3, which is designed for that exact purpose.
